Sources/SpeziHealthKit/SampleTypeScopedLocalStorage.swift

```diff
@@ -0,0 +1,103 @@
+            // this is fine bc we're not using it as a stable identifier
+            // (the `valueType` key must only be valid&unique for the lifetime of the app)
+            self.valueType = String(reflecting: valueType)
+            self.sampleType = sampleType.id
+        }
+    }
+    
+    private static let lock = NSLock()
+    nonisolated(unsafe) private static var localStorageKeys: [DictKey: Any] = [:]
+    
+    static func localStorageKey<Value>(
+        forValueType valueType: Value.Type,
+        sampleType: SampleType<some Any>,
+        defaultValue makeKey: @autoclosure () -> LocalStorageKey<Value>
+    ) -> LocalStorageKey<Value> {
+        lock.withLock {
+            let dictKey = DictKey(valueType: valueType, sampleType: sampleType)
+            if let key = localStorageKeys[dictKey] {
+                if let key = key as? LocalStorageKey<Value> {
+                    return key
+                } else {
+                    preconditionFailure("LocalStorageKey unexpectedly has incorrect type: expected '\(LocalStorageKey<Value>.self)'; got '\(type(of: key))'")
+                }
+            } else {
+                let key = makeKey()
+                localStorageKeys[dictKey] = key
+                return key
+            }
+        }
+    }
+}
+
+
+struct SampleTypeScopedLocalStorage<Value> {
+    private let localStorage: LocalStorage
+    private let makeStorageKey: @Sendable (any AnySampleType) -> LocalStorageKey<Value>
+    
+    private init(
+        localStorage: LocalStorage,
+        makeStorageKey: @escaping @Sendable (any AnySampleType) -> LocalStorageKey<Value>
+    ) {
+        self.localStorage = localStorage
+        self.makeStorageKey = makeStorageKey
+    }
+    
+    init(
+        localStorage: LocalStorage,
+        storageKeyPrefix: String,
+        storageSetting: LocalStorageSetting
+    ) where Value: NSObject & NSSecureCoding {
+        self.init(localStorage: localStorage) { sampleType in
+            LocalStorageKey<Value>("\(storageKeyPrefix).\(sampleType.id)", setting: storageSetting)
+        }
+    }
+    
+    init(
+        localStorage: LocalStorage,
+        storageKeyPrefix: String,
+        storageSetting: LocalStorageSetting
+    ) where Value: Codable {
+        self.init(localStorage: localStorage) { sampleType in
+            LocalStorageKey<Value>("\(storageKeyPrefix).\(sampleType.id)", setting: storageSetting)
+        }
+    }
+    
+    private func storageKey(for sampleType: SampleType<some Any>) -> LocalStorageKey<Value> {
+        LocalStorageKeysHandling.localStorageKey(forValueType: Value.self, sampleType: sampleType, defaultValue: makeStorageKey(sampleType))
+    }
+    
+    subscript(sampleType: SampleType<some Any>) -> Value? {
+        get {
+            try? localStorage.load(storageKey(for: sampleType))
+        }
+        nonmutating set {
+            try? localStorage.store(newValue, for: storageKey(for: sampleType))
+        }
```

Comment:
Is it smart to just return `nil` in case of an error here? Might make sense to make these getters / setters throwing, if the usage side then doesn't get too ugly.

I'm aware that this was previously done like that, just wondered when I read the code.

Reply:
we can make the getter throwing, but not the setter, and i wanted the API to be uniform across the two of them